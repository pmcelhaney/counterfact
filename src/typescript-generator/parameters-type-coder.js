import nodePath from "node:path";

import { Coder } from "./coder.js";
import { SchemaTypeCoder } from "./schema-type-coder.js";

export class ParametersTypeCoder extends Coder {
  constructor(requirement, placement) {
    super(requirement);

    this.placement = placement;
  }

  names() {
    return super.names("parameters");
  }

  write(script) {
    const typeDefinitions = this.requirement.data
      .filter((parameter) => parameter.in === this.placement)
      .map((parameter, index) => {
        const requirement = this.requirement.get(String(index));

        const { name, required } = parameter;

        const optionalFlag = required ? "" : "?";

        const schema = requirement.has("schema")
          ? requirement.get("schema")
          : requirement;

        return `"${name}${optionalFlag}": ${new SchemaTypeCoder(schema).write(
          script,
        )}`;
      });

    if (typeDefinitions.length === 0) {
      return "never";
    }

    return `{${typeDefinitions.join(", ")}}`;
  }

  modulePath() {
    const pathString = this.requirement.url
      .split("/")
      .at(-2)
      .replaceAll("~1", "/");

    return `${nodePath
      .join("parameters", pathString)
      .replaceAll("\\", "/")}.types.ts`;
  }
}
