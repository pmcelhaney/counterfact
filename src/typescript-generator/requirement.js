export class Requirement {
  constructor(data, url = "", specification = undefined) {
    this.data = data;
    this.url = url;
    this.specification = specification;
  }

  get isReference() {
    return this.data?.$ref !== undefined;
  }

  reference() {
    return this.specification.requirementAt(this.data.$ref, this.url);
  }

  has(item) {
    return item in this.data;
  }

  get(item) {
    if (!this.has(item)) {
      return undefined;
    }

    return new Requirement(
      this.data[item],
      `${this.url}/${item}`,
      this.specification,
    );
  }

  select(path, data = this.data, basePath = "") {
    const [head, ...tail] = path.split("/");

    const branch = data[this.unescapeJsonPointer(head)];

    if (!branch) {
      return undefined;
    }

    if (tail.length === 0) {
      return new Requirement(
        branch,
        `${this.url}/${basePath}${head}`,
        this.specification,
      );
    }

    return this.select(tail.join("/"), branch, `${basePath}${head}/`);
  }

  forEach(callback) {
    Object.keys(this.data).forEach((key) => {
      callback(this.select(this.escapeJsonPointer(key)), key);
    });
  }

  map(callback) {
    const result = [];

    // eslint-disable-next-line array-callback-return
    this.forEach((value, key) => result.push(callback(value, key)));

    return result;
  }

  flatMap(callback) {
    // eslint-disable-next-line unicorn/prefer-array-flat-map
    return this.map(callback).flat();
  }

  find(callback) {
    // eslint-disable-next-line init-declarations
    let result;

    this.forEach((value, key) => {
      // eslint-disable-next-line  n/callback-return
      if (result === undefined && callback(value, key)) {
        result = value;
      }
    });

    return result;
  }

  escapeJsonPointer(string) {
    return string.replaceAll("~", "~0").replaceAll("/", "~1");
  }

  unescapeJsonPointer(pointer) {
    return pointer.replaceAll("~1", "/").replaceAll("~0", "~");
  }
}
