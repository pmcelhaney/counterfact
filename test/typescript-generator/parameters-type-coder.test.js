import { describe, expect, it } from "@jest/globals";
import prettier from "prettier";

import { ParametersTypeCoder } from "../../src/typescript-generator/parameters-type-coder.js";
import { Requirement } from "../../src/typescript-generator/requirement.js";

function format(code) {
  return prettier.format(code, { parser: "typescript" });
}

describe("a ParametersTypeCoder", () => {
  it("generates a list of potential names", () => {
    const coder = new ParametersTypeCoder(new Requirement({}));

    const [one, two, three] = coder.names();

    expect([one, two, three]).toStrictEqual([
      "parameters",
      "parameters2",
      "parameters3",
    ]);
  });

  it("generates types for parameters", async () => {
    const requirement = new Requirement([
      { in: "path", name: "id", required: true, schema: { type: "string" } },
      { in: "query", name: "name", schema: { type: "string" } },
      { in: "query", name: "age", schema: { type: "number" } },
      { in: "header", name: "name", schema: { type: "string" } },
    ]);

    const pathCoder = new ParametersTypeCoder(requirement, "path");
    const queryCoder = new ParametersTypeCoder(requirement, "query");
    const headerCoder = new ParametersTypeCoder(requirement, "header");

    await expect(
      format(`type TestType =${pathCoder.write({})}`),
    ).resolves.toMatchSnapshot("path");
    await expect(
      format(`type TestType =${queryCoder.write({})}`),
    ).resolves.toMatchSnapshot("query");
    await expect(
      format(`type TestType =${headerCoder.write({})}`),
    ).resolves.toMatchSnapshot("header");
  });

  it("generates types for parameters (OAS2)", async () => {
    const requirement = new Requirement([
      { in: "path", name: "id", required: true, type: "string" },
      { in: "query", name: "name", type: "string" },
      { in: "query", name: "age", type: "number" },
      { in: "header", name: "name", type: "string" },
    ]);

    const pathCoder = new ParametersTypeCoder(requirement, "path");
    const queryCoder = new ParametersTypeCoder(requirement, "query");
    const headerCoder = new ParametersTypeCoder(requirement, "header");

    await expect(
      format(`type TestType =${pathCoder.write({})}`),
    ).resolves.toMatchSnapshot("path");
    await expect(
      format(`type TestType =${queryCoder.write({})}`),
    ).resolves.toMatchSnapshot("query");
    await expect(
      format(`type TestType =${headerCoder.write({})}`),
    ).resolves.toMatchSnapshot("header");
  });

  it("generates type never when there is no match", async () => {
    const requirement = new Requirement([
      { in: "path", name: "id", required: true, schema: { type: "string" } },
    ]);

    const coder = new ParametersTypeCoder(requirement, "query");

    await expect(
      format(`type TestType =${coder.write({})}`),
    ).resolves.toMatchSnapshot();
  });

  it("has no type", () => {
    const coder = new ParametersTypeCoder(new Requirement({}));

    expect(coder.typeDeclaration(undefined, {})).toBe("");
  });

  it("calculates the modulePath", () => {
    const coder = new ParametersTypeCoder(
      new Requirement({}, "/components/parameters/foo"),
    );

    expect(coder.modulePath()).toBe("parameters/parameters.types.ts");
  });
});
