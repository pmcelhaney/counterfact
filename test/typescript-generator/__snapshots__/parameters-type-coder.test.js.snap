// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`a ParametersTypeCoder generates type never when there is no match 1`] = `
"type TestType = never;
"
`;

exports[`a ParametersTypeCoder generates types for parameters (OAS2): header 1`] = `
"type TestType = { name?: string };
"
`;

exports[`a ParametersTypeCoder generates types for parameters (OAS2): path 1`] = `
"type TestType = { id: string };
"
`;

exports[`a ParametersTypeCoder generates types for parameters (OAS2): query 1`] = `
"type TestType = { name?: string; age?: number };
"
`;

exports[`a ParametersTypeCoder generates types for parameters: header 1`] = `
"type TestType = { name?: string };
"
`;

exports[`a ParametersTypeCoder generates types for parameters: path 1`] = `
"type TestType = { id: string };
"
`;

exports[`a ParametersTypeCoder generates types for parameters: query 1`] = `
"type TestType = { name?: string; age?: number };
"
`;
